src/transformer_sourcemap.ts

```diff
@@ -0,0 +1,144 @@
+
+export function createTransformerFromSourceMap(
+    operator: (sourceFile: ts.SourceFile, sourceMapper: SourceMapper) =>
+        string): ts.TransformerFactory<ts.SourceFile> {
+  return (context) => (sourceFile) => {
+    const sourceMapper = new NodeSourceMapper();
+    const newFile = ts.createSourceFile(
+        sourceFile.fileName, operator(sourceFile, sourceMapper), ts.ScriptTarget.Latest, true);
+    const mappedFile = visitNode(newFile);
+    return updateSourceFileNode(sourceFile, mappedFile.statements);
+
+    function visitNode<T extends ts.Node>(node: T): T {
+      return visitNodeWithSynthesizedComments(context, newFile, node, visitNodeImpl) as T;
+    }
+
+    function visitNodeImpl(node: ts.Node) {
+      if (node.flags & ts.NodeFlags.Synthesized) {
+        return node;
+      }
+      const originalNode = sourceMapper.getOriginalNode(node);
+
+      // Use the originalNode for:
+      // - literals: as e.g. typescript does not support synthetic regex literals
+      // - identifiers: as they don't have children and behave well
+      //    regarding comment synthesization
+      // - types: as they are not emited anyways
+      //          and it leads to errors with `extends` cases.
+      if (originalNode &&
+          (isLiteralKind(node.kind) || node.kind === ts.SyntaxKind.Identifier ||
+           isTypeNodeKind(node.kind) || node.kind === ts.SyntaxKind.IndexSignature)) {
+        return originalNode;
+      }
+      node = visitEachChild(node, visitNode, context);
+
+      node.flags |= ts.NodeFlags.Synthesized;
+      node.parent = undefined;
+      ts.setTextRange(node, originalNode ? originalNode : {pos: -1, end: -1});
+      ts.setOriginalNode(node, originalNode);
+
+      // tslint:disable-next-line:no-any
+      const nodeAny = node as {[key: string]: any};
+      // tslint:disable-next-line:no-any
+      const originalNodeAny = originalNode as {[key: string]: any};
+      for (const prop in nodeAny) {
+        if (nodeAny.hasOwnProperty(prop)) {
+          // tslint:disable-next-line:no-any
+          const value = nodeAny[prop];
+          if (isNodeArray(value)) {
+            // reset the pos/end of all NodeArrays so that we don't emit comments
+            // from them.
+            ts.setTextRange(value, {pos: -1, end: -1});
+          } else if (
+              isToken(value) && !(value.flags & ts.NodeFlags.Synthesized) &&
+              value.getSourceFile() !== sourceFile) {
+            // Use the original TextRange for all non visited tokens (e.g. the
+            // `BinaryExpression.operatorToken`) to preserve the formatting
+            const textRange = originalNode ? originalNodeAny[prop] : {pos: -1, end: -1};
+            ts.setTextRange(value, textRange);
+          }
+        }
+      }
+
+      return node;
+    }
+  };
+}
+
+class NodeSourceMapper implements SourceMapper {
```

Comment:
docs on here and the rest of this file

I think maybe this one could use a @fileOverview, because this is the one that does the clever "map node offsets to nodes" trick